includes/utils/class-amp-validation-utils.php

```diff
@@ -231,14 +301,133 @@ public static function validate_content( $post ) {
 		}
 	}
 
+	/**
+	 * Wraps callbacks in comments, to indicate to the sanitizer which plugin added them.
+	 *
+	 * Iterates through all of the registered callbacks.
+	 * If a callback is from a plugin and outputs markup,
+	 * this wraps the markup in comments.
+	 * Later, the sanitizer can identify which plugin any illegal markup is from.
+	 *
+	 * @global array $wp_filter
+	 * @return void
+	 */
+	public static function callback_wrappers() {
+		global $wp_filter;
+		if ( ! self::do_validate_front_end() ) {
+			return;
+		}
+		$pending_wrap_callbacks = array();
+		foreach ( $wp_filter as $filter_tag => $wp_hook ) {
+			foreach ( $wp_hook->callbacks as $priority => $callbacks ) {
+				foreach ( $callbacks as $callback ) {
+					$plugin = self::get_plugin( $callback['function'] );
+					if ( isset( $plugin ) ) {
+						$pending_wrap_callbacks[ $filter_tag ][] = array_merge(
+							$callback,
+							compact( 'plugin', 'priority' )
+						);
+					}
+				}
+			}
+		}
+
+		// Iterate over hooks to replace after iterating over all to begin with to prevent infinite loop in PHP<=5.4.
+		foreach ( $pending_wrap_callbacks as $hook => $callbacks ) {
+			foreach ( $callbacks as $callback ) {
+				remove_action( $hook, $callback['function'], $callback['priority'] );
+				$wrapped_callback = self::wrapped_callback( $callback );
+				add_action( $hook, $wrapped_callback, $callback['priority'], $callback['accepted_args'] );
+			}
+		}
+	}
+
+	/**
+	 * Gets the plugin of the callback, if one exists.
+	 *
+	 * @param string|array $callback The callback for which to get the plugin.
+	 * @return string|null $plugin   The plugin to which the callback belongs, or null.
+	 */
+	public static function get_plugin( $callback ) {
+		try {
+			if ( is_string( $callback ) && is_callable( $callback ) ) {
+				// The $callback is a function or static method.
+				$exploded_callback = explode( '::', $callback );
+				if ( count( $exploded_callback ) > 1 ) {
+					$reflection = new ReflectionClass( $exploded_callback[0] );
+				} else {
+					$reflection = new ReflectionFunction( $callback );
+				}
+			} elseif ( is_array( $callback ) && isset( $callback[0], $callback[1] ) && method_exists( $callback[0], $callback[1] ) ) {
+				// The $callback is a method.
+				$reflection = new ReflectionClass( $callback[0] );
+			} elseif ( is_object( $callback ) && ( 'Closure' === get_class( $callback ) ) ) {
+				$reflection = new ReflectionFunction( $callback );
+			}
+
+			$file = isset( $reflection ) ? $reflection->getFileName() : null;
+			if ( ! isset( $file ) ) {
+				return null;
+			}
+			$source_data = self::get_source( $file );
+			if ( 'plugins' === $source_data['type'] ) {
```

Comment:
What about `themes` and `mu-plugins`?

Reply:
This [now accommodates](https://github.com/Automattic/amp-wp/pull/971/files#diff-7d234989187d87a372e2aa6aa1299a71R349) themes and mu-plugins.